File: router/internal/blockhash/blockhash.go
package blockhash

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"hash"
)

const EmptyParent = "0000000000000000000000000000000000000000000000000000000000000000"

type IsolationKey struct {
	TenantID          string `json:"tenant_id"`
	TenantSalt        string `json:"tenant_salt"`
	ModelID           string `json:"model_id"`
	ModelRevision     string `json:"model_revision"`
	TokenizerRevision string `json:"tokenizer_revision"`
	LoRAID            string `json:"lora_id"`
	ModalityKey       string `json:"modality_key"`
	CacheSalt         string `json:"cache_salt"`
}

type Hasher struct {
	BlockSizeTokens int
}

func New(blockSizeTokens int) *Hasher {
	if blockSizeTokens <= 0 {
		blockSizeTokens = 16
	}
	return &Hasher{BlockSizeTokens: blockSizeTokens}
}

func (h *Hasher) ComputeBlocks(tokens []uint64, key IsolationKey) []string {
	if len(tokens) == 0 {
		return nil
	}
	parent := EmptyParent
	out := make([]string, 0, (len(tokens)+h.BlockSizeTokens-1)/h.BlockSizeTokens)
	for start := 0; start < len(tokens); start += h.BlockSizeTokens {
		end := start + h.BlockSizeTokens
		if end > len(tokens) {
			end = len(tokens)
		}
		hashHex := computeOne(parent, tokens[start:end], key)
		out = append(out, hashHex)
		parent = hashHex
	}
	return out
}

func computeOne(parent string, blockTokens []uint64, key IsolationKey) string {
	h := sha256.New()
	writeString(h, parent)
	writeString(h, key.TenantID)
	writeString(h, key.TenantSalt)
	writeString(h, key.ModelID)
	writeString(h, key.ModelRevision)
	writeString(h, key.TokenizerRevision)
	writeString(h, key.LoRAID)
	writeString(h, key.ModalityKey)
	writeString(h, key.CacheSalt)
	writeUint64(h, uint64(len(blockTokens)))
	for _, token := range blockTokens {
		writeUint64(h, token)
	}
	return hex.EncodeToString(h.Sum(nil))
}

func writeString(h hash.Hash, value string) {
	writeUint64(h, uint64(len(value)))
	_, _ = h.Write([]byte(value))
}

func writeUint64(h hash.Hash, value uint64) {
	var buf [8]byte
	binary.BigEndian.PutUint64(buf[:], value)
	_, _ = h.Write(buf[:])
}

File: shared/python/blockhash.py
from __future__ import annotations

import hashlib
import struct
from dataclasses import dataclass

EMPTY_PARENT = "0" * 64
FIELD_ORDER = (
    "tenant_id",
    "tenant_salt",
    "model_id",
    "model_revision",
    "tokenizer_revision",
    "lora_id",
    "modality_key",
    "cache_salt",
)


@dataclass(frozen=True)
class IsolationKey:
    tenant_id: str
    tenant_salt: str
    model_id: str
    model_revision: str
    tokenizer_revision: str
    lora_id: str = ""
    modality_key: str = ""
    cache_salt: str = ""

    @classmethod
    def from_dict(cls, data: dict) -> "IsolationKey":
        return cls(**{field: data.get(field, "") for field in FIELD_ORDER})


def _write_string(h: "hashlib._Hash", value: str) -> None:
    encoded = value.encode("utf-8")
    h.update(struct.pack(">Q", len(encoded)))
    h.update(encoded)


def _write_uint64(h: "hashlib._Hash", value: int) -> None:
    if value < 0:
        raise ValueError("token ids must be non-negative")
    h.update(struct.pack(">Q", value))


def compute_blocks(tokens: list[int], key: IsolationKey, block_size_tokens: int = 16) -> list[str]:
    if block_size_tokens <= 0:
        raise ValueError("block_size_tokens must be positive")
    parent = EMPTY_PARENT
    out: list[str] = []
    for start in range(0, len(tokens), block_size_tokens):
        block = tokens[start : start + block_size_tokens]
        h = hashlib.sha256()
        _write_string(h, parent)
        for field in FIELD_ORDER:
            _write_string(h, getattr(key, field))
        _write_uint64(h, len(block))
        for token in block:
            _write_uint64(h, int(token))
        parent = h.hexdigest()
        out.append(parent)
    return out
